### PR TITLE
[#90] improve error messages

### LINKED ISSUE #90: Improving the error messages from jesse when using oneOf/anyOf
When the JSON schema uses `oneOf`, `anyOf` or similar and the validation fails, the error message returned is not very helpful for the user. For example, in the case of `oneOf` the error is only `not_one_schema_valid`.

From what I've seen in the code, it should be possible to check all of the schemas in the specification provided in `anyOf`/`oneOf`, and add the errors from the schema with the least number of errors.

I am interested in implementing this change if you are willing to merge the PR 😄.

### PR DESCRIPTION
**Pull request by jfacorro**
### Description

Add internal errors for `oneOf`, `allOf` and `anyOf` to allow the user to figure out where the specific validation failed. 

Fixes #90.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -23,6 +23,9 @@ distclean:
 	$(RM) doc/stylesheet.css
 	$(RM) -r logs
 
+.PHONY: clean-tests
+clean-tests:
+	@ rm -rf _build/test/lib
```
Comment (jfacorro) [inline Makefile:28]: This is necessary to get it to work in Mac OS X.

File: src/jesse_schema_validator.hrl [hunk 2]
```diff
@@ -122,6 +122,7 @@
 -define(any_schemas_not_valid,       'any_schemas_not_valid').
 -define(not_multiple_of,             'not_multiple_of').
 -define(not_one_schema_valid,        'not_one_schema_valid').
+-define(more_than_one_schema_valid,  'more_than_one_schema_valid').
```
Comment (jfacorro) [inline src/jesse_schema_validator.hrl:125]: I thought adding this more specific error would be nice, but I can remove it if you think it's not worth it.

File: src/jesse_validator_draft4.erl [hunk 3]
```diff
@@ -1132,21 +1136,26 @@ check_all_of_(Value, [Schema | Schemas], State) ->
...
     {true, NewState} ->
-        case length(jesse_state:get_error_list(NewState)) of
-            NumErrsBefore -> NewState;
-            _  -> check_any_of_(Value, Schemas, State)
-        end;
-    {false, _} -> check_any_of_(Value, Schemas, State)
+      Errors = jesse_state:get_error_list(NewState),
+      case length(Errors) of
+        NumErrsBefore -> NewState;
+        _  -> check_any_of_(Value, Schemas, State, [Errors | SchemaErrors])
```
Comment (jfacorro) [inline src/jesse_validator_draft4.erl:1155]: I'm adding `Errors` in the list of `SchemaErrors` as returned from `jesse_state:get_error_list/1`. Not sure if I should extract the diff from the before and after list of errors returned by `jesse_state:get_error_list/1`.

Comment (seriyps) [reply src/jesse_validator_draft4.erl:1155]: Another option is to always keep only the smallest error: do smth like `check_any_of_(.., shortest_error(PreviousShortestError, NewErrors))`?

File: src/jesse_validator_draft4.erl [hunk 4]
```diff
@@ -69,7 +70,8 @@
                     | ?external.
 
 -type data_error_type() :: data_error()
-                         | {data_error(), binary()}.
+                         | {data_error(), binary()}
+                         | {data_error(), [jesse_error:error_reason()]}.
```
Comment (seriyps) [inline src/jesse_validator_draft4.erl:74]: I think `jesse_error:error_type()` needs to be updated too. (surprised it's not an alias).

`jesse_error:reason_to_jsx` might need an update as well, not sure.

File: Makefile (hunk 1 above)
Comment (jfacorro) [reply Makefile:28]: I tried some alternatives but this was the least hackish.

File: src/jesse_validator_draft4.erl (hunk 4 above)
Comment (jfacorro) [reply src/jesse_validator_draft4.erl:74]: Updated the `jesse_error:error_type()`, the function `jesse_error:reason_to_jsx/1` didn't need updating.

**Pull request merged**